src/buffer/out/TextAttribute.cpp

```diff
@@ -104,15 +104,34 @@ std::pair<COLORREF, COLORREF> TextAttribute::CalculateRgbColors(const std::array
                                                                 const COLORREF defaultBgColor,
                                                                 const bool reverseScreenMode,
                                                                 const bool blinkingIsFaint,
-                                                                const bool boldIsBright) const noexcept
+                                                                const bool boldIsBright,
+                                                                const std::optional<std::array<std::array<COLORREF, 18>, 18>>& adjustedForegroundColors) const noexcept
 {
-    auto fg = _foreground.GetColor(colorTable, defaultFgColor, boldIsBright && IsBold());
+    COLORREF fg;
     auto bg = _background.GetColor(colorTable, defaultBgColor);
+    bool reversed{ false };
+    if (adjustedForegroundColors.has_value() &&
+        (_background.IsDefault() || _background.IsLegacy()) &&
+        (_foreground.IsDefault() || _foreground.IsLegacy()))
+    {
+        auto bgIndex = _background.IsDefault() ? 16 : _background.GetIndex();
+        auto fgIndex = _foreground.IsDefault() ? 17 : _foreground.GetIndex();
+        fg = adjustedForegroundColors.value()[bgIndex][fgIndex];
+        if (IsReverseVideo() ^ reverseScreenMode)
+        {
+            std::swap(fg, bg);
+            reversed = true;

```

Comment:
there's probably a better way to ensure we don't re-reverse later rather than this flag

Reply:
What's the reason this early swap is needed here?

Reply:
Made a mistake here, the latest commit shows why we need the early swap:
If we do the swap later, we end up with
Background = adjustedForegroundColor based on initial bg/fg (pre-swap)
Foreground = Initial bg color

But what we want is
Background = Initial fg color
Foreground = adjustedForegroundColor based on new bg/fg (post-swap)